src/Microsoft.AspNetCore.Sockets.Client.Http/WebSocketsTransport.cs

```diff
@@ -132,77 +180,124 @@ public WebSocketsTransport(HttpOptions httpOptions, ILoggerFactory loggerFactory
+
+                    // Get a frame from the application
+
                     try
                     {
-                        if (!buffer.IsEmpty)
+                        if (result.IsCancelled)
                         {
-                            _logger.ReceivedFromApp(buffer.Length);
+                            break;
+                        }
 
-                            await _webSocket.SendAsync(buffer, webSocketMessageType, _transportCts.Token);
+                        if (!buffer.IsEmpty)
+                        {
+                            try
+                            {
+                                _logger.ReceivedFromApp(buffer.Length);
+
+                                if (WebSocketCanSend(ws))
+                                {
+                                    await ws.SendAsync(buffer, webSocketMessageType);
+                                }
+                                else
+                                {
+                                    break;
+                                }
+                            }
+                            catch (Exception ex)
+                            {
+                                if (!_aborted)
+                                {
+                                    _logger.ErrorSendingMessage(ex);
+                                }
+                                break;
+                            }
                         }
                         else if (result.IsCompleted)
                         {
                             break;
                         }
                     }
-                    catch (OperationCanceledException)
-                    {
-                        _logger.SendMessageCanceled();
-                        await CloseWebSocket();
-                        break;
-                    }
-                    catch (Exception ex)
-                    {
-                        _logger.ErrorSendingMessage(ex);
-                        await CloseWebSocket();
-                        throw;
-                    }
                     finally
                     {
                         _application.Input.AdvanceTo(buffer.End);
                     }
                 }
             }
-            catch (OperationCanceledException)
+            catch (Exception ex)
             {
-                _logger.SendCanceled();
+                error = ex;
             }
             finally
             {
-                _logger.SendStopped();
-                TriggerCancel();
+                if (WebSocketCanSend(ws))
+                {
+                    // We're done sending, send the close frame to the client if the websocket is still open
+                    await ws.CloseOutputAsync(error != null ? WebSocketCloseStatus.InternalServerError : WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
+                }
+
+                _application.Input.Complete();
```

Comment:
If `error != null` pass it into `Complete`?

Reply:
Input errors don't matter. Communicating them goes nowhere